src/mpid/ch4/shm/gpu/gpu_control.c

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+#include "gpu_pre.h"
+#include "gpu_impl.h"
+#include "gpu_recv.h"
+#include "gpu_control.h"
+
+int MPIDI_GPU_ctrl_send_ipc_recv_req_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_SHM_ctrl_gpu_send_recv_req_t *recv_req_hdr = &ctrl_hdr->gpu_recv_req;
+    MPIR_Request *rreq = NULL;
+    MPIR_Comm *root_comm;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_GPU_CTRL_SEND_IPC_RECV_REQ_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_GPU_CTRL_SEND_IPC_RECV_REQ_CB);
+
+    root_comm = MPIDIG_context_id_to_comm(recv_req_hdr->context_id);
+    if (root_comm) {
+        while (TRUE) {
+            rreq =
+                MPIDIG_dequeue_posted(recv_req_hdr->src_rank, recv_req_hdr->tag,
+                                      recv_req_hdr->context_id, 1, &MPIDIG_COMM(root_comm,
+                                                                                posted_list));
+#ifndef MPIDI_CH4_DIRECT_NETMOD
+            if (rreq) {
+                int is_cancelled;
+                mpi_errno = MPIDI_anysrc_try_cancel_partner(rreq, &is_cancelled);
+                MPIR_ERR_CHECK(mpi_errno);
+                if (!is_cancelled) {
+                    MPIR_Comm_release(root_comm);       /* -1 for posted_list */
+                    MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(rreq, datatype));
+                    continue;
+                }
+            }
+#endif
```

Comment:
Do we need this `ifndef` inside the shmmods?  This should not even be called when `MPIDI_CH4_DIRECT_NETMOD` is set, isn't it?

Reply:
Right none of this code should be called when using direct netmod. I have followed the template of XPMEM shmmod. I can remove the `ifndef`.

Reply:
I have removed the ifndef from this part of the code and added a stub for `MPIDI_anysrc_try_cancel_partner` in case of direct netmod is used.